realm/realm-library/src/main/java/io/realm/BaseRealm.java

```diff
@@ -86,6 +87,11 @@ public void onSchemaVersionChanged(long currentVersion) {
                             }
                         }, true);
         this.schema = new StandardRealmSchema(this);
+
+        // If waitForServerChanges() was enabled, we need to make sure that all data is downloaded
+        // before proceeding. We need to open the Realm instance first to start any potential underlying
+        // SyncSession so this will work. TODO: This needs to be decoupled.
+        ObjectServerFacade.getSyncFacadeIfPossible().downloadServerChangesIfNeeded(configuration);
```

Comment:
Currently all operations in the `RealmCache` is locked on the `RealmCache.class`, which means this will block opening all other different `RealmConfiguration`s. But never mind, i think having it here is OK, i will make the add a lock on individual `RealmCache` in the `getInstanceAsync()` PR.